File: navigation/common/src/test/java/androidx/navigation/NavGraphTest.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.navigation;

import static org.hamcrest.CoreMatchers.anyOf;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

import android.content.Context;
import android.support.annotation.IdRes;
import android.support.test.filters.SmallTest;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;

@SuppressWarnings("unchecked")
@RunWith(JUnit4.class)
@SmallTest
public class NavGraphTest {
    @IdRes
    private static final int FIRST_DESTINATION_ID = 1;
    @IdRes
    private static final int SECOND_DESTINATION_ID = 2;

    private NavGraphNavigator mNavGraphNavigator;

    @Before
    public void setup() {
        mNavGraphNavigator = new NavGraphNavigator(mock(Context.class));
    }

    private NavDestination createFirstDestination() {
        NavDestination destination = new NavDestination(mock(Navigator.class));
        destination.setId(FIRST_DESTINATION_ID);
        return destination;
    }

    private NavDestination createSecondDestination() {
        NavDestination destination = new NavDestination(mock(Navigator.class));
        destination.setId(SECOND_DESTINATION_ID);
        return destination;
    }

    private NavGraph createGraphWithDestination(NavDestination destination) {
        NavGraph graph = mNavGraphNavigator.createDestination();
        graph.addDestination(destination);
        return graph;
    }

    private NavGraph createGraphWithDestinations(NavDestination... destinations) {
        NavGraph graph = mNavGraphNavigator.createDestination();
        graph.addDestinations(destinations);
        return graph;
    }

    @Test(expected = IllegalArgumentException.class)
    public void addDestinationWithoutId() {
        NavGraph graph = mNavGraphNavigator.createDestination();
        NavDestination destination = new NavDestination(mock(Navigator.class));
        graph.addDestination(destination);
    }

    @Test
    public void addDestination() {
        NavDestination destination = createFirstDestination();
        NavGraph graph = createGraphWithDestination(destination);

        assertThat(destination.getParent(), is(graph));
        assertThat(graph.findNode(FIRST_DESTINATION_ID), is(destination));
    }

    @Test
    public void addDestinationsAsCollection() {
        NavGraph graph = mNavGraphNavigator.createDestination();
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        graph.addDestinations(Arrays.asList(destination, secondDestination));

        assertThat(destination.getParent(), is(graph));
        assertThat(graph.findNode(FIRST_DESTINATION_ID), is(destination));
        assertThat(secondDestination.getParent(), is(graph));
        assertThat(graph.findNode(SECOND_DESTINATION_ID), is(secondDestination));
    }

    @Test
    public void addDestinationsAsVarArgs() {
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        NavGraph graph = createGraphWithDestinations(destination, secondDestination);

        assertThat(destination.getParent(), is(graph));
        assertThat(graph.findNode(FIRST_DESTINATION_ID), is(destination));
        assertThat(secondDestination.getParent(), is(graph));
        assertThat(graph.findNode(SECOND_DESTINATION_ID), is(secondDestination));
    }

    @Test
    public void addReplacementDestination() {
        NavDestination destination = createFirstDestination();
        NavGraph graph = createGraphWithDestination(destination);

        NavDestination replacementDestination = new NavDestination(mock(Navigator.class));
        replacementDestination.setId(FIRST_DESTINATION_ID);
        graph.addDestination(replacementDestination);

        assertThat(destination.getParent(), nullValue());
        assertThat(replacementDestination.getParent(), is(graph));
        assertThat(graph.findNode(FIRST_DESTINATION_ID), is(replacementDestination));
    }

    @Test(expected = IllegalStateException.class)
    public void addDestinationWithExistingParent() {
        NavDestination destination = createFirstDestination();
        createGraphWithDestination(destination);

        NavGraph other = mNavGraphNavigator.createDestination();
        other.addDestination(destination);
    }

    @Test
    public void addAll() {
        NavDestination destination = createFirstDestination();
        NavGraph other = createGraphWithDestination(destination);

        NavGraph graph = mNavGraphNavigator.createDestination();
        graph.addAll(other);

        assertThat(destination.getParent(), is(graph));
        assertThat(graph.findNode(FIRST_DESTINATION_ID), is(destination));
        assertThat(other.findNode(FIRST_DESTINATION_ID), nullValue());
    }

    @Test
    public void removeDestination() {
        NavDestination destination = createFirstDestination();
        NavGraph graph = createGraphWithDestination(destination);

        graph.remove(destination);

        assertThat(destination.getParent(), nullValue());
        assertThat(graph.findNode(FIRST_DESTINATION_ID), nullValue());
    }

    @Test
    public void iterator() {
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        NavGraph graph = createGraphWithDestinations(destination, secondDestination);

        Iterator<NavDestination> iterator = graph.iterator();
        assertTrue(iterator.hasNext());
        assertThat(iterator.next(), anyOf(is(destination), is(secondDestination)));
        assertTrue(iterator.hasNext());
        assertThat(iterator.next(), anyOf(is(destination), is(secondDestination)));
        assertFalse(iterator.hasNext());
    }

    @Test(expected = NoSuchElementException.class)
    public void iteratorNoSuchElement() {
        NavDestination destination = createFirstDestination();
        NavGraph graph = createGraphWithDestination(destination);

        Iterator<NavDestination> iterator = graph.iterator();
        iterator.next();
        iterator.next();
    }

    @Test
    public void iteratorRemove() {
        NavDestination destination = createFirstDestination();
        NavGraph graph = createGraphWithDestination(destination);

        Iterator<NavDestination> iterator = graph.iterator();
        NavDestination value = iterator.next();
        iterator.remove();
        assertThat(value.getParent(), nullValue());
        assertThat(graph.findNode(value.getId()), nullValue());
    }

    @Test
    public void iteratorDoubleRemove() {
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        NavGraph graph = createGraphWithDestinations(destination, secondDestination);

        Iterator<NavDestination> iterator = graph.iterator();
        iterator.next();
        iterator.remove();
        NavDestination value = iterator.next();
        iterator.remove();
        assertThat(value.getParent(), nullValue());
        assertThat(graph.findNode(value.getId()), nullValue());
    }

    @Test(expected = IllegalStateException.class)
    public void iteratorDoubleRemoveWithoutNext() {
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        NavGraph graph = createGraphWithDestinations(destination, secondDestination);

        Iterator<NavDestination> iterator = graph.iterator();
        iterator.next();
        iterator.remove();
        iterator.remove();
    }

    @Test
    public void clear() {
        NavDestination destination = createFirstDestination();
        NavDestination secondDestination = createSecondDestination();
        NavGraph graph = createGraphWithDestinations(destination, secondDestination);

        graph.clear();
        assertThat(destination.getParent(), nullValue());
        assertThat(graph.findNode(FIRST_DESTINATION_ID), nullValue());
        assertThat(secondDestination.getParent(), nullValue());
        assertThat(graph.findNode(SECOND_DESTINATION_ID), nullValue());
    }
}

File: navigation/common/src/test/java/androidx/navigation/NavGraphTest.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.navigation

import android.content.Context
import android.support.annotation.IdRes
import androidx.test.filters.SmallTest
import com.google.common.truth.Truth.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mockito.mock
import java.util.Arrays
import java.util.NoSuchElementException

@RunWith(JUnit4::class)
@SmallTest
class NavGraphTest {

    companion object {
        @IdRes
        private const val FIRST_DESTINATION_ID = 1
        @IdRes
        private const val SECOND_DESTINATION_ID = 2
    }

    private lateinit var provider: NavigatorProvider
    private lateinit var noOpNavigator: NoOpNavigator
    private lateinit var navGraphNavigator: NavGraphNavigator

    @Before
    fun setup() {
        provider = NavigatorProvider().apply {
            addNavigator(NoOpNavigator().also { noOpNavigator = it })
            addNavigator(NavGraphNavigator(mock(Context::class.java)).also {
                navGraphNavigator = it
            })
        }
    }

    private fun createFirstDestination() = noOpNavigator.createDestination().apply {
        id = FIRST_DESTINATION_ID
    }

    private fun createSecondDestination() = noOpNavigator.createDestination().apply {
        id = SECOND_DESTINATION_ID
    }

    private fun createGraphWithDestination(destination: NavDestination) =
            navGraphNavigator.createDestination().apply {
                addDestination(destination)
            }

    private fun createGraphWithDestinations(vararg destinations: NavDestination) =
            navGraphNavigator.createDestination().apply {
                addDestinations(*destinations)
            }

    @Test(expected = IllegalArgumentException::class)
    fun addDestinationWithoutId() {
        val graph = navGraphNavigator.createDestination()
        val destination = noOpNavigator.createDestination()
        graph.addDestination(destination)
    }

    @Test
    fun addDestination() {
        val destination = createFirstDestination()
        val graph = createGraphWithDestination(destination)

        assertThat(destination.parent).isEqualTo(graph)
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isEqualTo(destination)
    }

    @Test
    fun addDestinationsAsCollection() {
        val graph = navGraphNavigator.createDestination()
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        graph.addDestinations(Arrays.asList(destination, secondDestination))

        assertThat(destination.parent).isEqualTo(graph)
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isEqualTo(destination)
        assertThat(secondDestination.parent).isEqualTo(graph)
        assertThat(graph.findNode(SECOND_DESTINATION_ID)).isEqualTo(secondDestination)
    }

    @Test
    fun addDestinationsAsVarArgs() {
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        val graph = createGraphWithDestinations(destination, secondDestination)

        assertThat(destination.parent).isEqualTo(graph)
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isEqualTo(destination)
        assertThat(secondDestination.parent).isEqualTo(graph)
        assertThat(graph.findNode(SECOND_DESTINATION_ID)).isEqualTo(secondDestination)
    }

    @Test
    fun addReplacementDestination() {
        val destination = createFirstDestination()
        val graph = createGraphWithDestination(destination)

        val replacementDestination = noOpNavigator.createDestination()
        replacementDestination.id = FIRST_DESTINATION_ID
        graph.addDestination(replacementDestination)

        assertThat(destination.parent).isNull()
        assertThat(replacementDestination.parent).isEqualTo(graph)
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isEqualTo(replacementDestination)
    }

    @Test(expected = IllegalStateException::class)
    fun addDestinationWithExistingParent() {
        val destination = createFirstDestination()
        createGraphWithDestination(destination)

        val other = navGraphNavigator.createDestination()
        other.addDestination(destination)
    }

    @Test
    fun addAll() {
        val destination = createFirstDestination()
        val other = createGraphWithDestination(destination)

        val graph = navGraphNavigator.createDestination()
        graph.addAll(other)

        assertThat(destination.parent).isEqualTo(graph)
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isEqualTo(destination)
        assertThat(other.findNode(FIRST_DESTINATION_ID)).isNull()
    }

    @Test
    fun removeDestination() {
        val destination = createFirstDestination()
        val graph = createGraphWithDestination(destination)

        graph.remove(destination)

        assertThat(destination.parent).isNull()
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isNull()
    }

    @Test
    operator fun iterator() {
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        val graph = createGraphWithDestinations(destination, secondDestination)

        assertThat(graph.iterator().asSequence().asIterable())
            .containsExactly(destination, secondDestination)
    }

    @Test(expected = NoSuchElementException::class)
    fun iteratorNoSuchElement() {
        val destination = createFirstDestination()
        val graph = createGraphWithDestination(destination)

        val iterator = graph.iterator()
        iterator.next()
        iterator.next()
    }

    @Test
    fun iteratorRemove() {
        val destination = createFirstDestination()
        val graph = createGraphWithDestination(destination)

        val iterator = graph.iterator()
        val value = iterator.next()
        iterator.remove()
        assertThat(value.parent).isNull()
        assertThat(graph.findNode(value.id)).isNull()
    }

    @Test
    fun iteratorDoubleRemove() {
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        val graph = createGraphWithDestinations(destination, secondDestination)

        val iterator = graph.iterator()
        iterator.next()
        iterator.remove()
        val value = iterator.next()
        iterator.remove()
        assertThat(value.parent).isNull()
        assertThat(graph.findNode(value.id)).isNull()
    }

    @Test(expected = IllegalStateException::class)
    fun iteratorDoubleRemoveWithoutNext() {
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        val graph = createGraphWithDestinations(destination, secondDestination)

        val iterator = graph.iterator()
        iterator.next()
        iterator.remove()
        iterator.remove()
    }

    @Test
    fun clear() {
        val destination = createFirstDestination()
        val secondDestination = createSecondDestination()
        val graph = createGraphWithDestinations(destination, secondDestination)

        graph.clear()
        assertThat(destination.parent).isNull()
        assertThat(graph.findNode(FIRST_DESTINATION_ID)).isNull()
        assertThat(secondDestination.parent).isNull()
        assertThat(graph.findNode(SECOND_DESTINATION_ID)).isNull()
    }
}
